GHA: Add clustermesh upgrade and downgrade tests

giorio94 opened the pull request:
This PR introduces a new GitHub actions workflow which validates the Cilium upgrade and downgrade paths when clustermesh is enabled, ensuring that this process does not disrupt long living connections.

More specifically, the test initially deploys a mixed version cluster mesh, with one cluster running the latest stable version of Cilium, and the other the current tip of main. It performs a subset of connectivity tests and deploys the application sensitive to connection disruption. At this point, the first cluster is upgraded to the tip of main, enabling kvstoremesh at the same time. Connectivity tests are again executed, additionally checking that no long living connection was dropped. Finally, the first cluster is downgraded again to the latest stable version of Cilium, disabling kvstoremesh. Connectivity tests and connection disruption checks are executed on more time.

To reduce the total amount of time required by this workflow, only a limited subset of tests is enabled, while the full suite is run in the conformance-clustermesh workflow.

Link to successful run: https://github.com/cilium/cilium/actions/runs/6379977851

giorio94 committed:
gha: add clustermesh upgrade and downgrade tests

This commit introduces a new GitHub actions workflow which validates the
Cilium upgrade and downgrade paths when clustermesh is enabled, ensuring
that this process does not disrupt long living connections.

More specifically, the test initially deploys a mixed version cluster
mesh, with one cluster running Cilium from the tip of the latest stable
branch, and the other the current tip of main. It performs a subset of
connectivity tests and deploys the application sensitive to connection
disruption.

At this point, the first cluster is upgraded to the tip of main,
enabling kvstoremesh at the same time. Connectivity tests are again
executed, checking that no long living connection was dropped. As an
additional stress test, the clustermesh-apiserver deployment in both
clusters is scaled to zero replicas, all agents restarted, and then
scaled back to one replica, while checking that no long living
connection was dropped.

Finally, the first cluster is downgraded again to the tip of the
latest stable version, disabling kvstoremesh. Connectivity tests and
connection disruption checks are executed one more time.

To reduce the total amount of time required by this workflow, only a
limited subset of tests is enabled, while the full suite is run in
the conformance-clustermesh workflow.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

aanm reviewed (approved):
LGTM. I'm pinging @brb since he is working on https://github.com/cilium/cilium-cli/pull/1970 which we could use to have longer and shorter connectivity tests in these upgrade tests depending if we are testing a PR or not.

giorio94 committed:
.github/actions/helm-default: make chart-dir configurable

Let's allow to override the default chart-dir, in case it is located
outside of the standard path. This can happen, for instance, when
checking out multiple Cilium versions in upgrade/downgrade workflows.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

aanm commented on .github/actions/helm-default/action.yaml:
```diff
@@ -35,7 +40,7 @@ runs:
           --helm-set=clustermesh.apiserver.image.tag=${SHA} \
           --helm-set=clustermesh.apiserver.image.useDigest=false \
-          --helm-set=clustermesh.apiserver.kvstoremesh.image.repository=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/clustermesh-apiserver-ci \
+          --helm-set=clustermesh.apiserver.kvstoremesh.image.repository=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/kvstoremesh-ci \
```
oh, nice catch!

aanm reviewed (approved):
Perfect
